File: playwright/candymapper/tests/Pages/CandyMapperPage.ts
import { expect, type Locator, type Page } from '@playwright/test';
import { BasePage } from './BasePage';

export class CandyMapperPage extends BasePage {
    // Add locators ghere
    private readonly PopupCloseIcon: Locator;
    private readonly PopupContainer: Locator;
    private readonly contactFirstName: Locator;
    private readonly contactLastName: Locator;
    private readonly contactEmail: Locator;
    private readonly contactPhone: Locator;
    private readonly contactMsg: Locator;
    private readonly submitButton: Locator;
    private readonly successMsg: Locator;
    private readonly errorMsg: Locator;
 
    constructor(page: Page) {
        super(page);
        this.PopupCloseIcon = page.locator('#popup-widget5912-close-icon');
        this.PopupContainer = page.locator('#popup-widget5912');
        this.contactFirstName = page.locator("input[id^='input'][data-aid='First Name']");
        this.contactLastName  = page.locator("input[id^='input'][data-aid='Last Name']");
        this.contactEmail     = page.locator("input[id^='input'][data-aid='CONTACT_FORM_EMAIL']");
        this.contactPhone     = page.locator("input[id^='input'][data-aid^='By entering a Phone Number']");
        this.contactMsg       = page.locator("textarea[placeholder='Message']");
        this.submitButton     = page.locator("button[type='submit']");
        this.successMsg       = page.locator("text=Thank you for your inquiry");
        this.errorMsg         = page.locator("text=Something went wrong");
    }
    
    async navigateCandyMapper(): Promise<this> {
        await this.navigateTo('https://candymapper.com/');
        return this;
    }

    async closePopup(): Promise<this> {
        await this.PopupContainer.waitFor({ state: 'visible', timeout: 15000 });
        await this.PopupCloseIcon.scrollIntoViewIfNeeded();
        await this.PopupCloseIcon.dispatchEvent('click');
        await this.PopupContainer.waitFor({ state: 'hidden', timeout: 10000 });
        return this;
    }

    async isPopupVisible(): Promise<boolean> {
        return this.PopupContainer.isVisible();
    }

    async completeField(field: 'last_name' | 'name' | 'email' | 'phone' | 'message'): Promise<this> {
    const fieldMap: Record<string, { locator: Locator; value: string }> = {
        last_name: { locator: this.contactLastName,  value: 'Robles' },
        name:      { locator: this.contactFirstName, value: 'Héctor' },
        email:     { locator: this.contactEmail,     value: 'hrobles@test.com' },
        phone:     { locator: this.contactPhone,     value: '5532514459' },
        message:   { locator: this.contactMsg,       value: 'Hello mate!' },
    };
    const { locator, value } = fieldMap[field];
    await this.safeType(locator, value);
    return this;
    }

    async clickButton(): Promise<Locator> {
        await this.submitButton.scrollIntoViewIfNeeded();
        await this.page.waitForTimeout(1000);
        
        await this.page.evaluate(() => {
            const badge = document.querySelector('.grecaptcha-badge') as HTMLElement;
            if (badge) badge.style.display = 'none';
        });
    
        // Back to native Playwright click
        //await this.page.screenshot({ path: 'before-click.png' });
        // First click focuses and triggers form validation JS
        await this.submitButton.click();
        await this.page.waitForTimeout(500);
        //await this.page.screenshot({ path: 'after-click.png' });
        // Second forced click submits - first click gets intercepted by page overlays
        await this.submitButton.click({force: true});
        console.log('Button clicked, waiting for response...');
    
        const result = await Promise.race([
            this.successMsg.waitFor({ state: 'visible', timeout: 20000 })
                .then(() => 'success' as const),
            this.errorMsg.waitFor({ state: 'visible', timeout: 20000 })
                .then(() => 'error' as const),
        ]);
    
        console.log(`Form submission result: ${result}`);
    
        if (result === 'error') {
            throw new Error('CandyMapper server error - retry the test');
        }
    
        return this.successMsg;
    }
}

File: selenium/pages/candymapper_page.py
from selenium.webdriver.common.by import By
from .base_page import BasePage
from selenium.webdriver.common.action_chains import ActionChains
import time

class CandyMapperPage(BasePage):
    #POPUP_CHALLENGE = (By.CSS_SELECTOR, "#popup-widget5912-close-icon path") # for JS option
    POPUP_CHALLENGE = (By.ID, "popup-widget5912-close-icon")
    POPUP_CONTAINER = (By.ID, "popup-widget5912")
    CONTACT_FIRSTNAME = (By.CSS_SELECTOR, "input[id^='input'][data-aid='First Name']")
    CONTACT_LASTNAME = (By.CSS_SELECTOR, "input[id^='input'][data-aid='Last Name']")
    CONTACT_EMAIL = (By.CSS_SELECTOR, "input[id^='input'][data-aid='CONTACT_FORM_EMAIL']")
    CONTACT_PHONE = (By.CSS_SELECTOR, "input[id^='input'][data-aid^='By entering a Phone Number']")
    CONTACT_MSG = (By.CSS_SELECTOR, "textarea[placeholder='Message']")
    SUBMIT_BUTTON = (By.XPATH, "//button[@type='submit']")
    SUCCESS_MSG_LOCATOR = (By.XPATH, "//span[contains(text(),'Thank you for your inquiry! We will get back to yo')]")
    ERROR_MSG_LOCATOR = (By.XPATH, "//p[@role='alertdialog']")

    test_input = {"name": "Héctor", "last_name": "Robles",
                  "phone": "5532514459", "email": "hrobles@test.com",
                  "message": "Hello mate!"}

    def navigate_candymapper(self):
        self.navigate_to("https://candymapper.com/")
        return self

    def wait_for_popup(self):
        self.wait_for_element(self.POPUP_CONTAINER, timeout=10)
        element = self.wait_for_clickable_element(self.POPUP_CHALLENGE, timeout=10)
        return element
    
    def is_popup_visible(self):
        return self.is_element_visible(self.POPUP_CONTAINER)

    def close_popup(self):
        element = self.wait_for_popup()
        # Define action chains since popup button is not a button but an SVG element
        ActionChains(self.driver).move_to_element(element).click().perform()
        # JavaScript option:
        #self.driver.execute_script("arguments[0].dispatchEvent(new MouseEvent('click', {bubbles: true}))", element)
        return self

    def find_textbox(self, locator):
        return self.driver.find_element(*locator)
    
    def complete_field(self, text_field):
        if 'last_name' in text_field:
            locator = self.CONTACT_LASTNAME
            value = self.test_input['last_name']  
        elif 'name' in text_field:
            locator = self.CONTACT_FIRSTNAME
            value = self.test_input['name']          
        elif 'email' in text_field:
            locator = self.CONTACT_EMAIL
            value = self.test_input['email']
        elif 'phone' in text_field:
            locator = self.CONTACT_PHONE
            value = self.test_input['phone']            
        elif 'message' in text_field:
            locator = self.CONTACT_MSG
            value = self.test_input['message']
        else: 
            raise ValueError(f"Unknown field: {text_field}")
        
        self._safe_send_keys(locator, value) # Tricky problem here eith page re-rendering
        # The issue here must have been caused by a javascript listener managing state
        # a Wait is added in order to let the elemenDOM stabilize before reaching any elem/form in this case
        return self.wait_for_clickable_element(locator, timeout=10).get_attribute("value")
        #assert actual == value, f"Field '{text_field}' not filled correctly: expected '{value}', got '{actual}'"
        #return self        

    def get_field_value(self, locator):
        return self.wait_for_clickable_element(locator).get_attribute("value")

    def click_button(self):
        element = self.wait_for_clickable_element(self.SUBMIT_BUTTON, timeout=10)
        self.driver.execute_script("arguments[0].scrollIntoView({block: 'center'});", element)
        self.driver.execute_script("document.querySelector('.grecaptcha-badge').style.display='none';")
        time.sleep(1)
        ActionChains(self.driver)\
            .move_to_element(element)\
            .pause(0.5)\
            .click()\
            .perform()
        return self.wait_for_element_presence(self.SUCCESS_MSG_LOCATOR, timeout=15)
    
    def find_mail_alert(self):
        return self.is_element_visible(self.ERROR_MSG_LOCATOR)